src/VisualStudio/Core/Def/Implementation/ProjectSystem/FileChangeWatcher.cs

```diff
@@ -133,26 +163,41 @@ private sealed class Context : IVsFreeThreadedFileChangeEvents2, IContext
             private readonly object _gate = new object();
             private bool _disposed = false;
             private readonly HashSet<FileWatchingToken> _activeFileWatchingTokens = new HashSet<FileWatchingToken>();
-            private uint _directoryWatchCookie;
 
-            public Context(FileChangeWatcher fileChangeWatcher, string? directoryFilePath)
+            /// <summary>
+            /// The list of cookies we used to make watchers for <see cref="_watchedDirectories"/>.
+            /// </summary>
+            /// <remarks>
+            /// This does not need to be used under <see cref="_gate"/>, as it's only used inside the actual queue of file watcher
+            /// actions.
+            /// </remarks>
+            private List<uint> _directoryWatchCookies = new List<uint>();
+
+            public Context(FileChangeWatcher fileChangeWatcher, ImmutableArray<WatchedDirectory> watchedDirectories)
             {
                 _fileChangeWatcher = fileChangeWatcher;
+                _watchedDirectories = watchedDirectories;
                 _noOpFileWatchingToken = new FileWatchingToken();
 
-                if (directoryFilePath != null)
+                foreach (var watchedDirectory in watchedDirectories)
                 {
-                    if (!directoryFilePath.EndsWith("\\"))
-                    {
-                        directoryFilePath += "\\";
-                    }
-
-                    _directoryFilePath = directoryFilePath;
-
                     _fileChangeWatcher.EnqueueWork(
                         async service =>
                         {
-                            _directoryWatchCookie = await service.AdviseDirChangeAsync(_directoryFilePath, watchSubdirectories: true, this).ConfigureAwait(false);
+                            var cookie = await service.AdviseDirChangeAsync(watchedDirectory.Path, watchSubdirectories: true, this).ConfigureAwait(false);
+                            _directoryWatchCookies.Add(cookie);
+
+                            if (watchedDirectory.ExtensionFilter != null)
+                            {
+                                // TODO: switch to proper reference assemblies
+                                var filterDirectoryChangesAsyncMethod = service.GetType().GetMethod("FilterDirectoryChangesAsync");
+
+                                if (filterDirectoryChangesAsyncMethod != null)
+                                {
+                                    var arguments = new object[] { cookie, new string[] { watchedDirectory.ExtensionFilter }, CancellationToken.None };
+                                    await ((Task)filterDirectoryChangesAsyncMethod.Invoke(service, arguments)).ConfigureAwait(false);
+                                }
+                            }
                         });
```

Comment:
so this work is async.  i don't really understand the lifetimes of this async work, esp. wrt to unadvising.

Reply:
We put all the actual calls to the service on a single queue, since the underlying VS service is more or less single threaded. Thus each operation here is added to a queue to be processed at a later time. The later unadvising still adds the operations to that queue.

What about the unadvising is confusing?